### ISSUE TITLE
`reportSelfClsParameterName` does not allow self param on callback protocol static methods

### ISSUE THREAD

**Issue by steverice** [microsoft/pyright#2121]
**Describe the bug**
In Pyright, it's unclear the "right" way to define a callback protocol for an instance method. The behavior of `reportSelfClsParameterName` seems to be most directly in the way of what would otherwise be the "best" solution in mypy + pyright.

**To Reproduce**
This is (a simplified version of) the code I want to describe by defining `MethodProtocol`:
```python
def decorator(func: MethodProtocol):
    return func

class MyClass:
    @decorator
    def my_method(self):
        pass
```

Each of the following options is problematic with pyright in some way:

These first two have obviously issues, and mypy errors on them as well.
```python
class MethodProtocol(Protocol):
    # pyright error: Keyword parameter "self" is missing in destination (reportGeneralTypeIssues)
    # mypy Argument 1 to "decorator" has incompatible type "Callable[[MyClass], Any]"; expected "MethodProtocol"
    def __call__(self): ...
```
```python
class MethodProtocol(Protocol):
    # pyright error: Duplicate parameter "self"
    # mypy error: Duplicate argument 'self' in function definition
    def __call__(self, self): ...
```

These equivalent solutions work under both pyright and mypy:
```python
class MethodProtocol(Protocol):
    @staticmethod
    def __call__(_self): ...

# or

class MethodProtocol(Protocol):
    def __call__(self, _self): ...
```

However no error is reported by pyright when defining
```python
class OtherClass:
    @decorator
    def my_invalid_method(other_arg):  # `other_arg` should have been `self`
        pass
```
Part of this is unfortunate, but expected — the argument name isn't being checked due to the leading underscore in `_self`.
But pyright also does not raise the `reportSelfClsParameterName` warning either.

Finally, in what seems like the "best" option,
```python
class MethodProtocol(Protocol):
    @staticmethod
    def __call__(self): ...
```
mypy raises no error, but pyright reports
```warning: Static methods should not take a "self" or "cls" parameter (reportSelfClsParameterName)```

**Expected behavior**
It's not clear what the "right" answer is here. In mypy, the best solution seems to be using `@staticmethod` in the Protocol, but under pyright this results in the `reportSelfClsParameterName` error.

I think an acceptable solution would be to suppress the `reportSelfClsParameterName` in this particular case. It certainly seems like an edge case, but I think a valid one. Turning off the `reportSelfClsParameterName` check to work around is a loss for code quality — it's genuinely useful in other circumstances.

**VS Code extension or command-line**
pyright CLI version 1.1.158

**Additional context**
Though the simple example I provide could also be accomplished with `Callable` instead of a callback protocol, the actual cases I'm looking at cannot be described without the protocol.

I tried to find a better alternative syntax for this construction but came up empty. As far as I'm aware I've presented the only ways to express this with Python types.

**Comment by erictraut** [comment]
Yeah, the way callback protocols were defined is unfortunate in that they require the extraneous `self` parameter. 

You have a few options here. First, you can use `_self` as the first parameter in the callback protocol, then use `self` as the second parameter. This works because pyright exempts parameters that start with underscores from the `reportSelfClsParameterName` check. This technique will flag an error with `my_invalid_method` in pyright, but mypy is a bit looser with its keyword parameter name consistency checks, so it doesn't emit an error here.

```python
class MethodProtocol(Protocol):
    def __call__(_self, self):
        ...
```

Second, you could specify in your callback protocol that the `self` parameter will never be specified using a keyword argument. To do this, insert a "position-only separator" after the second parameter. This will satisfy pyright and also work fine with mypy.

```python
class MethodProtocol(Protocol):
    def __call__(self, s, /):
        ...
```

**Comment by erictraut** [comment]
Third, you can use your `@staticmethod` technique but silence the `reportSelfClsParameterName` warning using a `# type: ignore` comment.

```python
class MethodProtocol(Protocol):
    @staticmethod
    def __call__(self):  # type: ignore
        ...
```

**Comment by steverice** [comment]
Thanks for the quick reply!

I like the position-only arguments option a lot but am using Python 3.7.
`def __call__(_self, self):` seems obvious in hindsight 🙂 

Both of those seem like acceptable solutions to me 💯

**Comment by erictraut** [comment]
If you're using Python 3.7, you can use the way of specifying a position-only parameter that predates PEP 570 — a parameter name that starts with a double-underscore. Both pyright and mypy honor this convention because it still appears in type stubs and some code bases.

```python
class MethodProtocol(Protocol):
    def __call__(self, __s):
        ...
```
